pkg/controller/pxc/upgrade.go

```diff
@@ -115,7 +110,208 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
 	currentSet.Spec.Template.Spec.Affinity = pxc.PodAffinity(podSpec.Affinity, sfs)
 	currentSet.Spec.Template.Spec.Volumes = sfsVolume.Volumes
 
-	return r.client.Update(context.TODO(), currentSet)
+	err = r.client.Update(context.TODO(), currentSet)
+	if err != nil {
+		return fmt.Errorf("update error: %v", err)
+	}
+
+	if cr.Spec.UpdateStrategy != v1.SmartUpdateStatefulSetStrategyType {
+		return nil
+	}
+
+	return r.smartUpdate(sfs, cr)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api.PerconaXtraDBCluster) error {
+	if !isPXC(sfs) {
+		return nil
+	}
+
+	if sfs.StatefulSet().Status.UpdatedReplicas >= sfs.StatefulSet().Status.Replicas {
+		return nil
+	}
+
+	log.Info("statefullSet was changed, run smart update")
+
+	if err := r.isBackupRunning(sfs.StatefulSet().Namespace); err != nil {
+		return fmt.Errorf("can't start 'SmartUpdate': %v", err)
+	}
+
+	if sfs.StatefulSet().Status.ReadyReplicas < sfs.StatefulSet().Status.Replicas {
+		return fmt.Errorf("can't start/continue 'SmartUpdate': waiting for all replicas are ready")
+	}
+
+	list := corev1.PodList{}
+	if err := r.client.List(context.TODO(),
+		&client.ListOptions{
+			Namespace:     sfs.StatefulSet().Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+		&list,
+	); err != nil {
+		return fmt.Errorf("get pod list: %v", err)
+	}
+
+	primary, err := r.getPrimaryPod(cr)
+	if err != nil {
+		return fmt.Errorf("get primary pod: %v", err)
+	}
+
+	log.Info(fmt.Sprintf("primary pod is %s", primary))
+
+	waitLimit := 120
+	if cr.Spec.PXC.LivenessInitialDelaySeconds != nil {
+		waitLimit = int(*cr.Spec.PXC.LivenessInitialDelaySeconds)
+	}
```

Comment:
I think we can move it from two places (here and in `pkg/pxc/app/statefulset/node.go`) into `CheckNSetDefaults` function and control default value in one place